blobstore/src/main/java/org/jclouds/blobstore/internal/BaseBlobStore.java

```diff
@@ -354,48 +350,22 @@ protected String putMultipartBlob(String container, Blob blob, PutOptions overri
    protected String putMultipartBlob(String container, Blob blob, PutOptions overrides, ListeningExecutorService executor) {
       ArrayList<ListenableFuture<MultipartPart>> parts = new ArrayList<ListenableFuture<MultipartPart>>();
       MultipartUpload mpu = initiateMultipartUpload(container, blob.getMetadata(), overrides);
-      // Cannot slice InputStream Payload since slice and close mutate the
-      // underlying stream.  Also issue synchronous uploads to avoid buffering
-      // arbitrary amounts of data in-memory.
       Payload payload = blob.getPayload();
-      boolean repeatable = blob.getPayload().isRepeatable();
-      if (!repeatable) {
-         payload = Payloads.newInputStreamPayload(new FilterInputStream((InputStream) payload.getRawContent()) {
-            @Override
-            public long skip(long offset) throws IOException {
-               // intentionally not implemented
-               return offset;
-            }
-
-            @Override
-            public void close() throws IOException {
-               // intentionally not implemented
-            }
-         });
-      }
-
-      try {
-         long contentLength = blob.getMetadata().getContentMetadata().getContentLength();
-         // TODO: inject MultipartUploadSlicingAlgorithm to override default part size
-         MultipartUploadSlicingAlgorithm algorithm = new MultipartUploadSlicingAlgorithm(
-               getMinimumMultipartPartSize(), getMaximumMultipartPartSize(), getMaximumNumberOfParts());
-         long partSize = algorithm.calculateChunkSize(contentLength);
-         int partNumber = 1;
-         while (partNumber <= algorithm.getParts()) {
-            Payload slice = slicer.slice(payload, algorithm.getCopied(), partSize);
+      boolean repeatable = payload.isRepeatable();
+      int partSize = (int) (getMinimumMultipartPartSize() + getMaximumMultipartPartSize()) / 2;
```

Comment:
For the common blobstore, S3, this will calculate (5 MB + 5 GB) / 2 = 2.5 GB.  In addition to overflowing an `int`, this is suboptimal for a lot of cases.  Most users are best served by several 5 MB parts uploading in parallel.  Can you devise a better algorithm that grows as the part size increases, e.g., parts 1-1000 are 5 MB, parts 2000-3000 are 25 MB, etc.?

Reply:
Thanks for pointing it out, I'll try to do as per your suggestion.